framework/src/main/java/org/checkerframework/framework/util/QualifierKindHierarchy.java

```diff
@@ -0,0 +1,839 @@
+
+        /** The annotation class for this. */
+        private final Class<? extends Annotation> clazz;
+
+        /** Whether or not this has elements. */
+        private final boolean hasElements;
+
+        /** Whether or not this is polymorphic. */
+        private boolean isPoly;
+
+        /**
+         * All the qualifier kinds that are a strict super qualifier kind of this. Does not include
+         * this qualifier kind itself.
+         */
+        private Set<QualifierKind> superTypes;
+
+        /** The top of the hierarchy to which this belongs. */
+        private QualifierKind top;
+
+        /** The bottom of the hierarchy to which this belongs. */
+        private QualifierKind bottom;
+
+        /**
+         * Creates a {@link QualifierKind} for the given annotation class.
+         *
+         * @param clazz annotation class for a qualifier
+         */
+        private QualifierKind(Class<? extends Annotation> clazz) {
+            this.clazz = clazz;
+            this.hasElements = clazz.getDeclaredMethods().length != 0;
+            this.name = clazz.getCanonicalName().intern();
+            // These non-final fields are set by QualifierKindHierarchy.
+            // Give them values here, because the Java compiler requires it.
+            isPoly = false;
+            superTypes = null;
+            top = null;
+            bottom = null;
+        }
+
+        /**
+         * Returns the canonical name of the annotation class of this.
+         *
+         * @return the canonical name of the annotation class of this
+         */
+        public String getName() {
+            return name;
+        }
+
+        /**
+         * Returns the annotation class for this.
+         *
+         * @return the annotation class for this
+         */
+        public Class<? extends Annotation> getAnnotationClass() {
+            return clazz;
+        }
+
+        /**
+         * Returns whether or not this is polymorphic.
+         *
+         * @return true if this is polymorphic
+         */
+        public boolean isPoly() {
+            return isPoly;
+        }
+
+        /**
+         * Whether or not this is the top qualifier of its hierarchy.
+         *
+         * @return true if this is the top qualifier of its hierarchy
+         */
+        public boolean isTop() {
+            return this.top == this;
+        }
+
+        /**
+         * Whether or not this is the bottom qualifier of its hierarchy.
+         *
+         * @return true if this is the bottom qualifier of its hierarchy
+         */
+        public boolean isBottom() {
+            return this.bottom == this;
+        }
+
+        /**
+         * All the qualifier kinds that are a strict super qualifier of this qualifier. Does not
+         * include this qualifier kind itself.
+         *
+         * @return all the qualifier kinds that are a strict super qualifier of this qualifier
+         */
+        public Set<QualifierKind> getSuperTypes() {
```

Comment:
It bugs me a bit that when we speak of subtypes, it is a reflexive operation, but when we speak of supertypes, it is strict (non-reflexive).  This lack of symmetry is confusing.  We could imagine renaming this method to `getStrictSuperTypes`.